### PR TITLE
APIs on per URL basis

### PR DESCRIPTION
**Pull request by vtapaskar**
Signed-off-by: Vijay Tapaskar <vijay@zededa.com>

- Doing APIs on per URL basis
- Starting to cleanup the hierarchy of structures

### CODE REVIEW

**Comment by deitch** [comment]
Is this intended to replace #64?

**Comment by deitch** [comment]
Also yetus checks are failing.

File: api/config/appconfig.proto [hunk 1]
```diff
@@ -3,15 +3,13 @@
 
 syntax = "proto3";
 
-option go_package  = "github.com/zededa/eve/sdk/go/zconfig";
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/appconfig.proto]: I think this is incorrect. The value of `go_package` is where go consuming packages will find this package. This is correct at `eve/sdk/go/config`, and not `api/config`. So, yes, change it from ending in `/zconfig` to ending in `/config`, but stays under `eve/sdk/go/`

File: api/config/baseosconfig.proto [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 syntax = "proto3";
 
-option go_package  = "github.com/zededa/eve/sdk/go/zconfig";
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/baseosconfig.proto]: Same as above

File: api/config/devcommon.proto [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Copyright(c) 2017-2018 Zededa, Inc.
+// All rights reserved.
+
+syntax = "proto3";
+
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/devcommon.proto]: Same as above: `github.com/zededa/eve/sdk/go/config`

File: api/config/devconfig.proto [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Copyright(c) 2017-2018 Zededa, Inc.
+// All rights reserved.
+
+syntax = "proto3";
+
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/devconfig.proto]: Same as above.

File: api/config/devmodel.proto [hunk 5]
```diff
@@ -0,0 +1,58 @@
+// Copyright(c) 2017-2018 Zededa, Inc.
+// All rights reserved.
+
+syntax = "proto3";
+
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/devmodel.proto]: Same as above: `github.com/zededa/eve/sdk/go/config`

File: api/config/fw.proto [hunk 6]
```diff
@@ -3,8 +3,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
 option java_package = "com.zededa.cloud.uservice.proto";
-option go_package = "github.com/zededa/eve/sdk/go/zconfig";
```
Comment (deitch) [inline api/config/fw.proto:7]: Same as above.

File: api/config/mesh.proto [hunk 7]
```diff
@@ -0,0 +1,26 @@
+// Copyright(c) 2017-2018 Zededa, Inc.
+// All rights reserved.
+
+syntax = "proto3";
+
+option go_package  = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/mesh.proto]: Same as above.

File: api/config/netcmn.proto [hunk 8]
```diff
@@ -3,8 +3,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
 option java_package = "com.zededa.cloud.uservice.proto";
-option go_package = "github.com/zededa/eve/sdk/go/zconfig";
```
Comment (deitch) [inline api/config/netcmn.proto:7]: Same as above.

File: api/config/netconfig.proto [hunk 9]
```diff
@@ -3,8 +3,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
```
Comment (deitch) [inline api/config/netconfig.proto]: Same as above.

File: api/config/netinst.proto [hunk 10]
```diff
@@ -3,8 +3,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
 option java_package = "com.zededa.cloud.uservice.proto";
-option go_package = "github.com/zededa/eve/sdk/go/zconfig";
```
Comment (deitch) [inline api/config/netinst.proto:7]: Same as above: `github.com/zededa/eve/sdk/go/config`

File: api/config/storage.proto [hunk 11]
```diff
@@ -4,8 +4,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
 option java_package = "com.zededa.cloud.uservice.proto";
-option go_package = "github.com/zededa/eve/sdk/go/zconfig";
```
Comment (deitch) [inline api/config/storage.proto:8]: Same as above.

File: api/config/vm.proto [hunk 12]
```diff
@@ -4,8 +4,8 @@
 
 syntax = "proto3";
 
+option go_package = "github.com/zededa/api/config";
 option java_package = "com.zededa.cloud.uservice.proto";
-option go_package = "github.com/zededa/eve/sdk/go/zconfig";
```
Comment (deitch) [inline api/config/vm.proto:8]: Same as above

File: api/logs/log.proto [hunk 13]
```diff
@@ -4,7 +4,7 @@
 syntax = "proto3";
 
 import "google/protobuf/timestamp.proto";
-option go_package  = "github.com/zededa/eve/sdk/go/zmet";
+option go_package  = "github.com/zededa/api/logs";
```
Comment (deitch) [inline api/logs/log.proto]: This should go to `github.com/zededa/eve/sdk/go/logs`

**Comment by rvs** [comment]
Hey @deitch -- lets move the discussion back to #64 since this is not standalone and for as long as it is WIP -- lets discuss it there. I'll push @zedvijay changes into #64 so we're all on the same page.

File: api/metrics/metrics.proto [hunk 14]
```diff
@@ -0,0 +1,296 @@
+// Copyright(c) 2017-2018 Zededa, Inc.
+// All rights reserved.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+option go_package  = "github.com/zededa/api/metrics";
```
Comment (deitch) [inline api/metrics/metrics.proto]: Same as above. This should be `github.com/zededa/eve/sdk/go/metrics`

File: api/register/register.proto [hunk 15]
```diff
@@ -3,7 +3,7 @@
 
 syntax = "proto3";
 
-option go_package  = "github.com/zededa/eve/sdk/go/zmet";
+option go_package  = "github.com/zededa/api/register";
```
Comment (deitch) [inline api/register/register.proto]: Same as above. This should be `github.com/zededa/eve/sdk/go/register`

**Review by deitch — CHANGES_REQUESTED** [review]
The `go_package` part is off. It should reflect not where this is, but rather where consumers will import it. We generate our sdk outputs to `github.com/zededa/eve/sdk/go/<package>`, so it should be there. Other than that, nice step ahead.

**Comment by deitch** [comment]
> lets move the discussion back to #64 since this is not standalone and for as long as it is WIP 

OK @rvs , although only saw this after doing a full review. Overall, it looks good, but the `go_package` parts are off. So let's not import this directly in there until those are fixed? Avoid double-work?

**Comment by zedvijay** [comment]
Thanks Avi, will fix those tomorrow am.

**Comment by deitch** [comment]
Cool @zedvijay !

**Comment by rvs** [comment]
Hey @zedvijay its not just what @deitch mentioned but also the questions that he asked on #64 

Since you guys didn't wait for me to update that - and we now have the split reviews between the two -- @deitch can you copy your questions into this thread? Some of them are important to be answered.

I'm closing #64 now

**Comment by rvs** [comment]
And one last thing @deitch any chance you can help update the docs for it (based on the answers to your questions + what's in the PR)?

**Comment by deitch** [comment]
From the other issue (edited), some questions:

1. Have we updated the consuming code in `pkg/pillar/` (and anywhere else in zedcloud)? It isn't strictly necessary if we versionize it properly, but we haven't done that yet, so we should update them in sync.
2. If we are removing the z prefix to the package, should we not also remove it from the messages? 
E.g.

```
  ZmNop = 0;
  ZmDevice = 1;
  ZmApp = 3;
}
```

should just be

```
  MNop = 0;
  MDevice = 1;
  MApp = 3;
}
```

My understanding is that there might be an issue with internal messages inside either zedcloud or `pkg/pillar/`, so I am not assuming this to be good.
To be clear, it also is OK to merge this with the `Z` in place, and then remove those (if we can) as a separate PR.

3. Once confirmed, immediately after merging in or as part of this PR, we should update the API document.

**Comment by deitch** [comment]
> And one last thing @deitch any chance you can help update the docs for it 

Definitely. Let's get this ready to run first.

**Comment by zedvijay** [comment]
Avi,

Please inline.

Vijay

> On May 7, 2019, at 11:39 PM, Avi Deitcher <notifications@github.com> wrote:
> 
> From the other issue (edited), some questions:
> 
> Have we updated the consuming code in pkg/pillar/ (and anywhere else in zedcloud)? It isn't strictly necessary if we versionize it properly, but we haven't done that yet, so we should update them in sync.
This is not done yet. It will be follow up commit. Controller side of changes are already done, we should start changing the pillar code.

> If we are removing the z prefix to the package, should we not also remove it from the messages?
> E.g.
>   ZmNop = 0;
>   ZmDevice = 1;
>   ZmApp = 3;
> }
> should just be
> 
>   MNop = 0;
>   MDevice = 1;
>   MApp = 3;
> }
> My understanding is that there might be an issue with internal messages inside either zedcloud or pkg/pillar/, so I am not assuming this to be good.
> To be clear, it also is OK to merge this with the Z in place, and then remove those (if we can) as a separate PR.
> 
> 

100% agree. Also I would like to add documentation around what each of these things means inline. Thats next thing on my plate.

> Once confirmed, immediately after merging in or as part of this PR, we should update the API document.
Either way is fine by me.

**Comment by rvs** [comment]
I'll need @zedvijay to provide answers to some of these, but here's my take:
   * we haven't updated what's vendored in pillar yet -- that'll be a separate follow up PR (based on the vendoring mechanics I describe in the next bullet item)
   * we're no longer vendoring api explicitly because it is done at the pillar level (based on the go.mod)
   * yes it would be nice to get rid of z-prefix, but I'll leave it up to @zedvijay to decide if it is too much of a brain surgery at this point

Makes sense?

**Comment by deitch** [comment]
> This is not done yet. It will be follow up commit. Controller side of changes are already done, we should start changing the pillar code.

Sounds good. According to what Roman said above, we are versioning it in with go modules. As long as we are tying it to specific commit, we are good. These changes affect other commits when run to `sdk/go/`, so we have the flexibility to do it in sequential commits. Nice.

> If we are removing the z prefix to the package, should we not also remove it from the messages? E.g. ZmNop = 0; ZmDevice = 1; ZmApp = 3; } should just be MNop = 0; MDevice = 1; MApp = 3; } My understanding is that there might be an issue with internal messages inside either zedcloud or pkg/pillar/, so I am not assuming this to be good. To be clear, it also is OK to merge this with the Z in place, and then remove those (if we can) as a separate PR.
> 100% agree. Also I would like to add documentation around what each of these things means inline. Thats next thing on my plate.

That would be appreciated. Even as I wrote the API doc/spec with everyone's help, some of the specific messages were a little opaque to me.

Also @rvs wrote:

> Makes sense?

Yes, indeed. All on the same page. Sounds like we should:

1. Fix the `go_package` issue
2. Fix whatever is causing `yetus` to fail tests
3. Merge this one in
4. Update the docs as a separate PR
5. Update pillar as a separate PR

**Review by deitch — APPROVED** [review]

**Comment by rvs** [comment]
This is back to you @zedvijay it seems ;-)

**Comment by deitch** [comment]
It would have been better not to have merged this in when CI was failing. Granted, it was just yetus, but we need to be consistent about it going forward.

**Pull request merged**